docs/source/external.rst

```diff
@@ -0,0 +1,297 @@
+             'temperature': 5.0}
+    'descriptor': '219310e0-faa0-4990-84a0-95b508d4ae35',
+    ...}
+
+now that we know that ``'image'`` is external, the value
+``'aa10035d-1d2b-41d9-97e6-03e3fe62fa6c/5'`` must be a ``datum_id``,
+referencing a Datum document. Here is the matching Datum document. This
+document can be used to retrieve some data that belong in our Event. In our
+example it might be an image or stack of images that were taken at a
+given temperature during a temperature scan.
+
+.. code:: python
+
+   # 'datum' document
+   {'datum_id': 'aa10035d-1d2b-41d9-97e6-03e3fe62fa6c/5',
+    'datum_kwargs': {'index': 5},
+    'resource': 'aa10035d-1d2b-41d9-97e6-03e3fe62fa6c'}
+
+You will notice that we still do not have a *filepath* anywhere here. It is
+common for many Datum documents to point into the same file (e.g. a large HDF5
+file) or series of files (e.g. TIFF series).  Rather than store that
+information separately and redundantly in each Datum, the Datum documents point
+to a Resource document---the last document type we'll need here---which
+contains path-related details.
+
+ .. code:: python
+
+   # 'resource' document
+   {'uid': 'aa10035d-1d2b-41d9-97e6-03e3fe62fa6c',
+    'spec': 'AD_HDF5',
+    'root': '/GPFS/DATA/Andor/',
+    'resource_path': '2020/01/03/8ff08ff9-a2bf-48c3-8ff3-dcac0f309d7d.h5',
+    'resource_kwargs': {'frame_per_point': 10},
+    'path_semantics': 'posix',
+    'uid': '3b300e6f-b431-4750-a635-5630d15c81a8',
+    'run_start': '10bf6945-4afd-43ca-af36-6ad8f3540bcd'}
+
+The ``resource_path`` is a relative path, all of which is semantic and should
+usually not change during the lifecycle of this asset. The ``root`` is more
+context-dependent (depending on what system you are accessing the data from)
+and subject to change (if the data is moved over time).
+
+The ``spec`` gives us a hint about the format of this asset, whether it be a
+file, multiple files, or something more specialized. The ``resource_kwargs``
+provide any additional parameters for reading it.
+
+Handlers
+========
+
+In bluesky/databroker, a "handler" is a reader with a special interface. It
+accepts a Resource document and a Datum document and in exchange returns the
+pertinent data.
+
+Handler Interface
+-----------------
+
+A 'handler class' may be any callable with the signature::
+
+    handler_class(full_path, **resource_kwargs)
+
+It is expected to return an object, a 'handler instance', which is also
+callable and has the following signature::
+
+    handler_instance(**datum_kwargs)
+
+As the names 'handler class' and 'handler instance' suggest, this is
+typically implemented using a class that implements ``__init__`` and
+``__call__``, with the respective signatures.
+
+.. code:: python
+
+   class MyHandler:
+       def __init__(self, path, **resource_kwargs):
+           # Consume the path information and the 'resource_kwargs' from the
+           # Resource. Typically stashes some state and/or opens file(s).
+           ...
```

Comment:
I wonder if it would help to describe what the expected lifecycle of a handler is? Does DB cache the handler instance for consecutive reads? How does that cache expire?

Reply:
That is an implementation detail we have struggled with and one of the things that motivated the design of RunRouter to scope the life to "between the start and stop documents" which is commented on later.

IRRC, the cache is injectable so it is also possible to have very custom policies on a per-application basis.

Reply:
Seconded. The Filler and RunRouter classes make it easier to clean up correctly and set custom cache policies, but I still feel we haven’t settled on what the right policies should be for DataBroker.

There are some comments on resource management later on in the document. I will add some more here.

Reply:
Added comments in https://github.com/bluesky/event-model/pull/136/commits/cf4d0097085b435fb52acf1aa0e8ee890aca16e4. Any thoughts?

Reply:
Very helpful explanation; the flexibility is interesting. Thanks!